src/bgw/job.c

```diff
@@ -176,84 +180,168 @@ ts_bgw_job_get_all(size_t alloc_size, MemoryContext mctx)
+ *  modified while in the middle of a run. Note that in regular DML operations,
+ *  heavyweight locks on tuples are only sometimes taken as otherwise the MVCC
+ *  system uses tuple headers for its locking. Thus, we need to take this lock
+ *  explicitly ourselves during DML and treat this more as an advisory lock system
+ *  managed by ourselves.
+ *
+ * returns whether or not the lock was obtained (false return only possible if dontwait==false)
+ */
 
-	ScanKeyInit(&scankey[0],
-				Anum_bgw_job_stat_pkey_idx_job_id,
-				BTEqualStrategyNumber,
-				F_INT4EQ,
-				Int32GetDatum(bgw_job_id));
-	bgw_job_scan_one(BGW_JOB_STAT_PKEY_IDX,
-					 scankey,
-					 1,
-					 tuple_found,
-					 tuple_filter,
-					 data,
-					 mctx,
-					 lockmode,
-					 fail_if_not_found);
+static bool
+lock_tuple(HeapTuple heap, Relation relation, LOCKMODE mode, bool session_lock, LOCKTAG *tag,
+		   bool block)
+{
+	ItemPointer tid = &heap->t_self;
+	SET_LOCKTAG_TUPLE(*tag,
+					  relation->rd_lockInfo.lockRelId.dbId,
+					  relation->rd_lockInfo.lockRelId.relId,
+					  ItemPointerGetBlockNumber(tid),
+					  ItemPointerGetOffsetNumber(tid));
+
+	return LockAcquire(tag, mode, session_lock, !block) != LOCKACQUIRE_NOT_AVAIL;
 }
 
-static ScanTupleResult
-bgw_job_tuple_found(TupleInfo *ti, void *const data)
+static BgwJob *
+ts_bgw_job_find_with_lock(int32 bgw_job_id, MemoryContext mctx, LOCKMODE tuple_lock_mode,
+						  LOCKTAG *session_lock_locktag, bool block)
 {
-	BgwJob **job_pp = data;
+	/* Take a share lock on the table to prevent concurrent data changes during scan. This lock will
+	 * be released after the scan */
+	ScanIterator iterator = ts_scan_iterator_create(BGW_JOB, ShareLock, mctx);
+	int num_found = 0;
+	BgwJob *job = NULL;
+	LOCKTAG tmp;
+	LOCKTAG *tag = session_lock_locktag;
 
-	*job_pp = bgw_job_from_tuple(ti->tuple, sizeof(BgwJob), ti->mctx);
+	if (tag == NULL)
+		tag = &tmp;
 
-	/*
-	 * Return SCAN_CONTINUE because we check for multiple tuples as an error
-	 * condition.
-	 */
-	return SCAN_CONTINUE;
+	init_scan_by_job_id(&iterator, bgw_job_id);
+
+	ts_scanner_foreach(&iterator)
+	{
+		HeapTuple heap = ts_scan_iterator_tuple(&iterator);
+		job = bgw_job_from_tuple(heap, sizeof(BgwJob), mctx);
+		if (!lock_tuple(heap,
+						iterator.ictx.tablerel,
+						tuple_lock_mode,
+						session_lock_locktag != NULL,
+						tag,
+						block))
+		{
+			/* return NULL if lock could not be acquired */
+			Assert(!block);
+			pfree(job);
+			job = NULL;
+		}
+
+		Assert(num_found == 0);
+		num_found++;
+	}
+
+	return job;
+}
+
+/* Take a share lock on the job tuple ofr the duration of the session. This prevents
+ *  the job from being deleted.
+ *
+ *  Returns whether or not the job still exists.
+ */
+bool
+ts_bgw_job_get_share_lock(int32 bgw_job_id, MemoryContext mctx)
+{
+	/* note the mode here is equivalent to FOR SHARE row locks */
+	BgwJob *job = ts_bgw_job_find_with_lock(bgw_job_id, mctx, RowShareLock, NULL, true);
```

Comment:
The comment above says it take the lock for the session, but the session lock is NULL, is this correct?

Reply:
the comments was incorrect. fixed